EDAM limitations update (#4)
bioexcel/cwl-best-practice-guide: docs/limitations.md

It is recommended, where possible, to use a format definition taken from a widely-used ontological resource, such as [EDAM](http://edamontology.org). Doing so will enable users to reference the formats that your workflow expects, so that they can ensure they match these.

Note, however, that the cwltool currently compares the format string only, it does not check the file for conformance to that format definition. It is up to you to make sure that you use the correct definition for your file format. If a format definition for your files is not available in the existing ontological databases, then you can set your own format string - however it is recommended that you do this in your own namespace, so it is clear it is not part of any existing database. It is also possible to request the addition of your format to a database, e.g. by raising an [issue for EDAM on github](https://github.com/edamontology/edamontology/issues).

## Managing file names and file storage

> CWL does ensure separation of working space for each step within a workflow, however output files cannot be renamed as they are gathered from steps, and the requested outputs are returned to the base directory from which the workflow is run. This can cause issues when using `scatter` to parallise a workflow, as outputs can potentially overwrite each other. It also can result in a messy workflow directory, if a lot of outputs are being returned.

Reviewer: >  however output files cannot be renamed as they are gathered from steps

Sure, but they can be autonamed as part of the underlying `CommandLineTool` itself

> as outputs can potentially overwrite each other

That would be non-compliant with the CWL standards. Please remove this statement.

>  It also can result in a messy workflow directory
 
There is no concept in CWL of a `workflow directory`

Author: Thanks for the review :)

My comments here are influenced by the issues I had with organising the outputs from a workflow built on top of the BioBB building blocks (this is the main workflow file: [md_list.cwl](https://github.com/douglowe/biobb_hpc_cwl_md_list/blob/main/md_list.cwl)).

Their tool descriptors do not autoname the output files returned from the tool, so I had to write an extra tool to gather these outputs and place them into a directory with a unique name for each instance of the workflow. Otherwise the returned files would overwrite each other. The files were all returned to me in the same directory that I ran the workflow from - which is what I mean by `workflow directory`. Is there a way for a CWL user to specify a different directory for each `scatter` instance? I couldn't see anything that would do this when I was developing this workflow, but perhaps I missed the option somewhere?

Reviewer: From our conversation yesterday: the results of a scatter are always kept in a different directory in the CWL reference runner and derivatives (as is required by the CWL standards). There was a bug for a while that when writing the final outputs some files with the same file name were clobbered; this should be fixed, so please test again and let us know which runner still has that problem.

Either way, please change the text of this section. 
```suggestion
CWL does ensure separation of working space for each step within a workflow, however output files cannot be renamed as they are gathered from steps. It can result in a messy final workflow output directory for some systems, if a lot of outputs are being returned.
```

Author: That makes sense to me. I'll correct this text now, and will test cwl-runner & toil to see how they deal with this kind of situation.

Author: This has been done in [PR 30](https://github.com/bioexcel/cwl-best-practice-guide/pull/30).

> One solution for this is to add a final step to your workflow that gathers together all desired outputs into a single directory. This can most easily be performed using an `ExpressionTool` to create a uniquely named directory for each scatter instance containing the required files. This will need the  `InlineJavascriptRequirement` requirement though, and so will mean your workflow is less portable than it would otherwise be.

Reviewer: >  and so will mean your workflow is less portable than it would otherwise be.

Basically all CWL engines support `InlineJavascriptRequirement`, though it is technically an optional part (like supporting `Workflow`s at all :-)

So I would appreciate it if this statement was removed

Author: Yep - happy to modify/remove this statement. I was thinking of the requirement of installing a java engine to be able to support these tools, which I had issues with when first learning CWL. So I could perhaps reword this too:
"This will need the `InlineJavascriptRequirement` requirement though, and to support using this you will need to install java (using, for example, the conda-forge `nodejs` package) or docker (enabling CWL to access java support via the `openjdk` docker image)."

Reviewer: "This will need the InlineJavascriptRequirement requirement though, and to support using this with `cwltool` or `toil-cwl-runner` you will need to install the Node javascript engine (using, for example, the conda-forge `nodejs` package) or docker (enabling CWL to access javascript support via the nodejs docker image)."

Author: I've opened a new [PR](https://github.com/bioexcel/cwl-best-practice-guide/pull/29) for this correction - could you have a look at that and let me know if it looks okay?
